Ltc2992 add initial support

tachicialex opened the pull request:
Add initial support for ltc2992. 
Currently exposes only current and voltage on SENSE1 and SENSE2 and 
voltages on the 4 GPIO inputs via the onboard VADC.

commodo commented on drivers/hwmon/ltc2992.c:
```diff
@@ -0,0 +1,263 @@
+		if (addr > 1)
+			return -EINVAL;
+
+		ret = fwnode_property_read_u32(child, "shunt-resistor-micro-ohms", &val);
```
from what i see, some drivers do a default value of 1000 micro-ohms
should we do something similar for this one as well?

commodo replied:
i think it's something like:
```
if (fwnode_property_read_u32(child, "shunt-resistor-micro-ohms", &val))
    val = 1000;

st->supplies[addr].r_sense_uohm = val;
```

if going for a 1000 micro-ohm default

commodo replied:
i see now:
```
			if (st->supplies[channel].r_sense_uohm)
				return 0444;
```

commodo commented on drivers/hwmon/ltc2992.c:
```diff
@@ -0,0 +1,263 @@
+			if (channel > 1)
+				return 0444;
+
+			if (st->supplies[channel].r_sense_uohm)
```
is zero a valid value?
maybe we shouldn't allow this at probe?

commodo commented on drivers/hwmon/ltc2992.c:
```diff
@@ -0,0 +1,263 @@
+
+	hwmon_dev = devm_hwmon_device_register_with_info(&client->dev, client->name, st,
+							 &ltc2992_chip_info, NULL);
+	if (IS_ERR(hwmon_dev))
```
will there be more stuff here?
if not, we could do  `return PTR_ERR_OR_ZERO(hwmon_dev)`   directly here

commodo commented on drivers/hwmon/ltc2992.c:
```diff
@@ -0,0 +1,263 @@
+
+static struct i2c_driver ltc2992_i2c_driver = {
+	.driver = {
+		.name = "ltc2992",
```
i'd also add an of_table here; this allows also probing via OF and ACPI;

nunojsa replied:
Moreover, I would actually only use the `of_table` since there's really no way that this will probe without devicetree

commodo commented on drivers/hwmon/ltc2992.c:
```diff
@@ -0,0 +1,263 @@
+	st->client = client;
+	st->regmap = devm_regmap_init_i2c(client, &ltc2992_regmap_config);
+	if (IS_ERR(st->regmap)) {
+		dev_err(&client->dev, "devm_regmap_init_i2c failed!\n");
```
i think maybe remove this error message;
upstream will consider it spam

commodo commented on drivers/hwmon/ltc2992.c:
```diff
@@ -0,0 +1,263 @@
+		}
+		break;
+	default:
+		break;
```
`return -EOPNOTSUPP` here? or -EINVAL or what is the correct return here

commodo commented on drivers/hwmon/ltc2992.c:
```diff
@@ -0,0 +1,263 @@
+			/* get voltage of GPIOs */
+			if (channel > 1) {
+				nr_gpio = channel - 2;
+				ret = ltc2992_read_reg(st, ltc2992_gpio_addr_map[nr_gpio], 2, &reg);
```
maybe do error checks for `ltc2992_read_reg()` returns?

commodo commented on drivers/hwmon/ltc2992.c:
```diff
@@ -0,0 +1,263 @@
+			/* get voltage of SENSE+ */
+			ret = ltc2992_read_reg(st, LTC2992_SENSE(channel), 2, &reg);
+			reg = mul_u64_u32_div(reg, LTC2992_VADC_MAX_MILLIV, LTC2992_ADCS_SCALE);
+			break;
```
maybe add some default cases to these internal switch() statements?

commodo commented on drivers/hwmon/ltc2992.c:
```diff
@@ -0,0 +1,263 @@
+		regvals[reg_len - i - 1] = (val >> (i * 8)) & 0xFF;
+	}
+
+	ret = regmap_bulk_write(st->regmap, addr, regvals, reg_len);
```
`return regmap_bulk_write()` directly?

commodo commented on drivers/hwmon/ltc2992.c:
```diff
@@ -0,0 +1,263 @@
+
+	*val = 0;
+	for (i = 0; i < reg_len; i++)
+		*val += regvals[reg_len - i - 1] << (i * 8);
```
this looks correct;
just be aware that some upstream people may want `*val |=`  instead of `*val +=`
it just looks more familiar

commodo commented on drivers/hwmon/ltc2992.c:
```diff
@@ -0,0 +1,263 @@
+		case hwmon_curr_input:
+			ret = ltc2992_read_reg(st, LTC2992_DSENSE(channel), 2, &reg);
+			reg = mul_u64_u32_div(reg, LTC2992_IADC_MAX_UV, LTC2992_ADCS_SCALE);
+			reg = mul_u64_u32_div(reg, 1000, st->supplies[channel].r_sense_uohm);
```
would it be an idea to do all multiplications first, then all the divisions?
this would [probably] have a slightly better precision during integer division

commodo commented on drivers/hwmon/ltc2992.c:
```diff
@@ -0,0 +1,263 @@
+	return ret;
+}
+
+static int ltc2992_write(struct device *dev, enum hwmon_sensor_types type, u32 attr, int channel,
```
i'm a bit un-familiar with hwmon;
is this dummy write() going to be implemented later? or is it required by the hwmon framework?

nunojsa replied:
It depends if you define your attrs as writable or not. In this case, it seems you are defining them as read only which means that you should not really need the `write()` callback. If you are thinking about in adding it the future, I would add the callback by then...

nunojsa commented on drivers/hwmon/ltc2992.c:
```diff
@@ -0,0 +1,263 @@
+
+	fwnode = dev_fwnode(&st->client->dev);
+
+	fwnode_for_each_child_node(fwnode, child) {
```
`fwnode_get_next_available_child_node()` instead?

commodo commented on Documentation/devicetree/bindings/hwmon/adi,ltc2992.yaml:
```diff
@@ -0,0 +1,74 @@
+         minimum: 0
+         maximum: 1
+
+      shunt-resistor-micro-ohms:
```
it isn't; it's common to many hwmon drivers;
it's the sense resistor value for converting volts to amps

nunojsa replied:
I see. I was wondering if we need the vendor prefix here

commodo replied:
i don't think we do; at some point, this could be moved to hwmon core;
i also have to implement this support for ltc2945

commodo commented:
also, this patchset requires a commit for building the driver through the CI

tachicialex commented on drivers/hwmon/ltc2992.c:
```diff
@@ -0,0 +1,263 @@
+#define LTC2992_ADCS_SCALE		65536
+
+struct ltc2992_supply {
+	u32				r_sense_uohm;
```
Yes, min, max thresholds for alarm in next patch.

commodo commented on drivers/hwmon/ltc2992.c:
```diff
@@ -0,0 +1,263 @@
+	LTC2992_G1, LTC2992_G2, LTC2992_G3, LTC2992_G4
+};
+
+static int ltc2992_read_reg(struct ltc2992_state *st, u8 addr, u8 reg_len, u32 *val)
```
these `const int` things are not [yet] very common in kernel code;
when upstreaming a driver i would keep this as close to common kernel practice;

i would leave it as AlexT did here [for now];
otherwise, this gets sent to upstream and we hear back that they don't like `const int` [for whatever reason] and then all this was just internal review noise;

there's really no benefit in const-ifying an int; since it's a copy on the stack

nunojsa replied:
Well, I would argue that using `const` here is a good practice since it helps readability and is self documenting things. Plus, it allows the compiler to do some assumptions (even though I guess the compiler is smart enough to figure that out without the `const` qualifier and the performance gain should be minimal...). Still, I think it makes the API more clear and I try always to remember to make things `const` when they are. Still no complains from upstream guys because of this :smile: .

Either way, this is a nit so...

tachicialex commented:
Changelog v1 -> v2:
- hwmon_in_input - voltage can always be provided even if there is no shunt r
- changed from if (IS_ERR(hwmon_dev)) ret -> return PTR_ERR_OR_ZERO(hwmon_dev)
- added of_table
- removed regmap init error message
- added checks for ltc2992_read_reg
- removed hwmon dummy write for now
- for hwmon_curr_input calculations multiply everything first then divide
- make shunt-resistor-micro-ohms optional as there is no need for it in order for the device/driver
to function normally. Sense+ and Sense- can be shorted. In this way only the voltage of Sense+
will be exposed to userspace and not the current.
- added commit for selecting LTC2992 in the hwmon ADI Kconfig

nunojsa commented on drivers/hwmon/ltc2992.c:
```diff
@@ -0,0 +1,260 @@
+			long *val)
+{
+	struct ltc2992_state *st = dev_get_drvdata(dev);
+	int ret = -EOPNOTSUPP;
```
nit: you are using `ENOTSUPP` in the rest of the function. Why using `EOPNOTSUPP` here? I would say to either use one or the other

tachicialex commented:
Changelog V3:
- applied EOPNOTSUPP everywhere instead of ENOTSUPP.

commodo commented on Documentation/devicetree/bindings/hwmon/adi,ltc2992.yaml:
```diff
@@ -0,0 +1,74 @@
+required:
+  - compatible
+  - reg
+
```
add a line here `additionalProperties: false`
it seems this is now required in new DTs

commodo replied:
actually i am not sure now; you have `patternProperties` defined; so no idea if `additionalProperties: false` is needed;

commodo commented on Documentation/devicetree/bindings/hwmon/adi,ltc2992.yaml:
```diff
@@ -0,0 +1,74 @@
+  reg:
+    maxItems: 1
+
+  avcc-supply:
```
looking at other schemas; this should only be `avcc-supply: true`

commodo commented on drivers/hwmon/ltc2992.c:
```diff
@@ -0,0 +1,260 @@
+
+MODULE_AUTHOR("Alexandru Tachici <alexandru.tachici@analog.com>");
+MODULE_DESCRIPTION("Hwmon driver for Linear Technology 2992");
+MODULE_LICENSE("GPL v2");
```
new requirement that has appeared internally, is to make Linux drivers Dual license [with BSD]
this is so that we can let customers copy Linux drivers for FreeBSD [and friends]

let's make this: `MODULE_LICENSE("Dual BSD/GPL");`
this means that `// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)`

or should we do: `// SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)`
@mhennerich 
i have no idea what the difference between  BSD-2-Clause  &   BSD-3-Clause

tachicialex commented:
Changelog v3 -> v4:
- replaced fwnode_for_each_child_node with
fwnode_for_each_available_child_node
- replaced GPL2.0 licence with GPL-2.0-only OR BSD-2-Clause)
- changed avcc-supply: to avcc-supply: true
- added in yaml:   additionalProperties: false
- added in yaml: req properties for child nodes

commodo commented on Documentation/devicetree/bindings/hwmon/adi,ltc2992.yaml:
```diff
@@ -0,0 +1,79 @@
+    - reg
+    - shunt-resistor-micro-ohms
+
+  additionalProperties: false
```
this looks duplicate?
or is it needed?

tachicialex replied:
Apparently patternProperties require that too. Saw this on multiple upstream yaml files.
